Fix metafunction for many pointwise operations

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113635
* #113674
* __->__ #113634

The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc @ezyang @eellison @bdhirsh @mruberry @Lezcano @peterbell10

lezcano committed:
Fix metafunction for many pointwise operations

The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

[ghstack-poisoned]

peterbell10 commented on torch/testing/_internal/opinfo/definitions/special.py:
```diff
@@ -61,6 +62,21 @@ def sample_inputs_i0_i1(op_info, device, dtype, requires_grad, **kwargs):
+    if dtype == torch.float32:
+        yield SampleInput(
+            make_arg((S,), dtype=dtype), make_arg((L, 1), dtype=torch.float16)
+        )
```
This seems a bit hacky. Would be better as an opinfo test that checks one sample with different dtype inputs.

lezcano replied:
Fair enuf

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -3575,8 +3599,8 @@ def meta_binop_inplace_alpha(self, other, alpha=1):
-        self, type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT
+    return elementwise_meta(
+        self, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
```
For unary ops, was prims meta not good enough?

lezcano replied:
I guess so, but I figured it'd be good to have just one source of truth, rather than using a private function from a different module. Everything is a no-op when you have just one tensor, so that should be alright.

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -3600,8 +3624,8 @@ def shift_dtype_check(fn_name, self, val):
-    element_wise = _elementwise_meta(
-        self, type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT
+    element_wise = elementwise_meta(
+        self, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
```
This still looks fishy as we aren't broadcasting or doing type promotion with `other`.

lezcano replied:
Yeah, that doesn't seem right. Let me fix that

peterbell10 replied:
This hasn't been addressed

lezcano replied:
ugh, sorry, the rebase messed it up. I'll fix it

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -51,6 +52,29 @@ def register(op):
+    )
+    result = _prim_elementwise_meta(*args, type_promotion=type_promotion_prim)
+
+    return _maybe_convert_to_dtype(result, result_dtype)
```
Does this ever do anything? Shouldn't `_prim_elementwise_meta` have done the type conversions already?

lezcano replied:
It does. Consider that the inputs are `bfloat16`. Then the computation dtype is `float32`, but hte result_dtype is `bfloat16`.

peterbell10 replied:
We shouldn't be using computation dtype for meta functions though

lezcano replied:
I thought about that, but I was not sure whether the two are commutative.
I guess I could use `result_type` directly + `ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT` unconditionally.

lezcano committed:
Update on "Fix metafunction for many pointwise operations"


The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc ezyang eellison bdhirsh mruberry Lezcano peterbell10

[ghstack-poisoned]

lezcano commented:
ghstack has not updated this PR, although it tells me locally that it has... :/

lezcano committed:
Update on "Fix metafunction for many pointwise operations"


The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc ezyang eellison bdhirsh mruberry Lezcano peterbell10

[ghstack-poisoned]

peterbell10 commented:
@lezcano I think your changes have ended up in the wrong PRs.

lezcano committed:
Update on "Fix metafunction for many pointwise operations"


The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc ezyang eellison bdhirsh mruberry Lezcano peterbell10

[ghstack-poisoned]

lezcano commented:
Yes they had. Fixed now

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -3614,8 +3635,8 @@ def meta_rshifts(self, other):
     shift_dtype_check("lshift", self, other)
-    element_wise = _elementwise_meta(
-        self, type_promotion=ELEMENTWISE_PRIM_TYPE_PROMOTION_KIND.DEFAULT
+    element_wise = elementwise_meta(
+        self, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
     # Annoying edgecase
     if self.dim() == 0 and isinstance(other, torch.Tensor):
```
Same issue here
```suggestion
    element_wise = elementwise_meta(
        self, other, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
    )
```

lezcano committed:
Update on "Fix metafunction for many pointwise operations"


The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc ezyang eellison bdhirsh mruberry Lezcano peterbell10

[ghstack-poisoned]

lezcano commented:
hopefully now the whole stack should be ready

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -3600,8 +3621,8 @@ def shift_dtype_check(fn_name, self, val):
+    element_wise = elementwise_meta(
+        self, other, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
     # Annoying edgecase
     if self.dim() == 0 and isinstance(other, torch.Tensor):
```
```suggestion
```

peterbell10 commented on torch/_meta_registrations.py:
```diff
@@ -3614,8 +3635,8 @@ def meta_rshifts(self, other):
+    element_wise = elementwise_meta(
+        self, other, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
     # Annoying edgecase
     if self.dim() == 0 and isinstance(other, torch.Tensor):
```
```suggestion
```

lezcano committed:
Update on "Fix metafunction for many pointwise operations"


The previous metafunction was completely broken.
It incorrectly used a metafunction that was designed for prims. It also
passed in an incorrect enum class for the type promotion.

Fixes https://github.com/pytorch/pytorch/issues/113119

cc ezyang eellison bdhirsh mruberry Lezcano peterbell10

[ghstack-poisoned]
